server/src/main/java/io/crate/analyze/Id.java

```diff
@@ -28,67 +28,86 @@
-        return ensureNonNull(getOnlyElement(keyValues));
+    private static final BiFunction<List<String>, List<DataType<?>>, String> ONLY_ITEM_NULL_VALIDATION = (strings, dataTypes) -> {
+        DataType<?> type = Lists.getOnlyElement(dataTypes);
+        String value = ensureNonNull(Lists.getOnlyElement(strings));
+        if (type.id() == DataTypes.IP.id()) {
+            return InetAddresses.forString(value).toString();
+        }
+        return value;
     };
 
-    private static final Function<List<String>, String> ONLY_ITEM = Lists::getOnlyElement;
-
+    private static final BiFunction<List<String>, List<DataType<?>>, String> ONLY_ITEM = (strings, dataTypes) -> {
+        DataType<?> type = Lists.getOnlyElement(dataTypes);
+        String value = Lists.getOnlyElement(strings);
+        if (type.id() == DataTypes.IP.id()) {
+            return InetAddresses.forString(value).toString();
+        }
+        return value;
+    };
 
     /**
      * generates a function which can be used to generate an id and apply null validation.
      * <p>
      * This variant doesn't handle the pk = _id case.
      */
-    private static Function<List<String>, String> compileWithNullValidation(final int numPks,
-                                                                            final int clusteredByPosition) {
-        return switch (numPks) {
-            case 0 -> RANDOM_ID;
-            case 1 -> ONLY_ITEM_NULL_VALIDATION;
-            default -> keyValues -> {
-                if (keyValues.size() != numPks) {
-                    throw new IllegalArgumentException("Missing primary key values");
-                }
-                return encode(keyValues, clusteredByPosition);
-            };
-        };
-    }
 
     /**
      * generates a function which can be used to generate an id.
      * <p>
      * This variant doesn't handle the pk = _id case.
      */
-    public static Function<List<String>, String> compile(final int numPks, final int clusteredByPosition) {
+    public static BiFunction<List<String>, List<DataType<?>>, String> compile(final int numPks, final int clusteredByPosition) {
         if (numPks == 1) {
             return ONLY_ITEM;
         }
-        return compileWithNullValidation(numPks, clusteredByPosition);
+        return switch (numPks) {
+            case 0 -> RANDOM_ID;
+            case 1 -> ONLY_ITEM_NULL_VALIDATION;
+            default -> (keyValues, pkDataTypes) -> {
+                if (keyValues.size() != numPks) {
+                    throw new IllegalArgumentException("Missing primary key values");
+                }
+                List<String> values = new ArrayList<>(keyValues.size());
+                for (int i = 0; i < keyValues.size(); i++) {
+                    String value = keyValues.get(i);
+                    DataType<?> type = pkDataTypes.get(i);
+                    if (type.id() == DataTypes.IP.id()) {
```

Comment:
This should be some method on the dataType that does this transparently, without the need of the `if`.